File: python/fasta.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" This file is a simple python script reading a fast file and save it
to memory into Fasta object.
"""

import os


class Fasta(object):
    """ Class representing a FASTA sequence.
    This class has two attributes, a name corresponding to the header
    line and a data corresponding to the sequence itself.
    """

    def __init__(self, name=None, data=None):
        """ Constructor. """
        self.name = name
        self.data = data

    def __repr__(self):
        return '<Fasta "%s">' % self.name


def read_fasta(filename):
    """ Read a fasta file and return a list of Fasta object. """
    if not os.path.exists(filename):
        print "No file '%s' found" % filename
        return
    stream = open(filename, 'r')
    fasta = None
    fasta_seq = []
    for row in stream.readlines():
        if row.startswith('>'):
            if fasta:
                fasta_seq.append(fasta)
            name = row.split('>', 1)[1].strip()
            fasta = Fasta(name = name)
        else:
            if fasta is not None:
                if fasta.data:
                    fasta.data = fasta.data + row.strip()
                else:
                    fasta.data = row.strip()
    if fasta:
        fasta_seq.append(fasta)
    return fasta_seq


if __name__ == '__main__':
    for seq in read_fasta(os.path.join(os.path.dirname(
        os.path.abspath( __file__ )),
        '..', 'sample.fasta')):
        print seq.name, len(seq.data)

File: cpp/fasta.cpp
#include "fasta.h"
#include <string.h>
#include <stdio.h> // provides printf

const char* Fasta::header()
{
    return strdup(_header.c_str());
}

const char* Fasta::data()
{
    return strdup(_data.c_str());
}

int Fasta::length()
{
    return _data.length();
}

int Fasta::gccontent()
{
    if (_data.length() == 0){
        return 0;
    }
    int gccontent = 0;

    std::string s(_data);
    for (std::string::iterator it = s.begin(), end = s.end(); it != end; ++it)
    {
        char c = tolower(*it);
        if (c == 'c' || c == 'g')
        {
            gccontent = gccontent + 1;
        }
    }
    return (gccontent * 100 / _data.length());
}

void Fasta::setName(const std::string& name)
{
    _header = name;
}

void Fasta::setSequence(const std::string& sequence)
{
    _data = sequence;
}

void Fasta::addSequence(const std::string& sequence)
{
    if (!_data.empty())
    {
        _data = _data + sequence;
    } else {
        _data = sequence;
    }
}
